### PR TITLE
Aggressive retry

### PR DESCRIPTION
**Pull request by rohandubal**
*Description of changes:*

Allow developers to specify a desired retry strategy. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: AWSAppSyncClient/AWSAppSyncClientConfiguration.swift [hunk 1]
```diff
@@ -365,6 +384,8 @@ public class AWSAppSyncClientConfiguration {
                                                                urlSessionConfiguration: urlSessionConfiguration,
                                                                authProvider: oidcAuthProvider)
         }
+        
+        networkTransport.retryStrategy = retryStrategy
```
Comment (palpatim) [inline AWSAppSyncClient/AWSAppSyncClientConfiguration.swift]: pass the retry strategy as part of `makeNetworkTransport...` method family to keep all the assignments together

File: AWSAppSyncClient/AWSAppSyncHTTPNetworkTransport.swift [hunk 2]
```diff
@@ -18,6 +18,9 @@ public class AWSAppSyncHTTPNetworkTransport: AWSNetworkTransport {
     var oidcAuthProvider: AWSOIDCAuthProvider? = nil
     var endpoint: AWSEndpoint? = nil
     let authType: AWSAppSyncAuthType
+    // We use the default strategy as exponential retry if someone is using AWSAppSyncHTTPTransport directly.
+    // Currently, retry strategy cannot be changed individually for HTTP network; it is controlled at client level.
+    var retryStrategy: AWSAppSyncRetryStrategy = .exponential
```
Comment (palpatim) [inline AWSAppSyncClient/AWSAppSyncHTTPNetworkTransport.swift]: Add this as an initialization parameter with a default

File: AWSAppSyncClient/Internal/AWSAppSyncRetryHandler.swift [hunk 3]
```diff
@@ -12,12 +12,25 @@ struct AWSAppSyncRetryAdvice {
...
+// Current implementation has heavy coupling of retry strategy, appsync config and http transport.
+// The logic of shouldRetryRequest should be decoupled for different retryStrategies.
+// TODO: Implement a protocol which accepts a standardized set of inputs, errorResponse, attemptNumber, etc.
+// and returns if retry should be done and after what duration.
+// We can also extend the `AWSAppSyncRetryStrategy` to accept a `custom` enum type which contains a class
+// implementing above protocol.
 final class AWSAppSyncRetryHandler {
     static let maxWaitMilliseconds = 300 * 1000 // 5 minutes of max retry duration.
-
+    static let maxRetryAttemptsWhenUsingAggresiveMode = 12
```
Comment (palpatim) [inline AWSAppSyncClient/Internal/AWSAppSyncRetryHandler.swift]: In the interests of making things more self contained & testable, what about having actual implementations separately? Something like `AggressiveRetryStrategy` and `ExponentialBackoffRetryStrategy`, which have only one entry point: `retryDelayInMillseconds(for attemptNumber: Int) -> Int`. All logic and configurations can be localized there, and it keeps this file more readable against a hypothetical future where we add another strategy ("laid back"? "hyper aggressive?" :) )

File: AWSAppSyncUnitTests/AWSAppSyncRetryHandlerTests.swift [hunk 4]
```diff
@@ -144,4 +144,56 @@ class AWSAppSyncRetryHandlerTests: XCTestCase {
...
+            }
+            XCTAssertTrue(1000 ... 1099 ~= retryAdvice.retryInterval!.intValueForMillis()!)
+            print("retryAdvice: should retry in \(retryAdvice.retryInterval ?? .seconds(-1))")
+        }
+        XCTAssertFalse(retryAdvice.shouldRetry)
+    }
+    
+    func test_exponentialRetryDuration() {
+        let expAttempt2 = AWSAppSyncRetryHandler.retryDelayInMillseconds(for: 2, retryStrategy: .exponential)
+        XCTAssert(400 ... 499 ~= expAttempt2, "\(expAttempt2) value is out of expected range.")
```
Comment (palpatim) [inline AWSAppSyncUnitTests/AWSAppSyncRetryHandlerTests.swift]: +1 for pattern matches!

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,16 @@
 
 The AWS AppSync SDK for iOS enables you to access your AWS AppSync backend and perform operations like `Queries`, `Mutations` and `Subscriptions`. The SDK also includes support for offline operations.
 
+## 2.10.4
+
+### New Features
+
+- The AppSyncClient now supports specifying retry strategy using the `retryStrategy` parameter in `AWSAppSyncClientConfiguration`. You can choose between `aggressive` and `exponential`(default selection).
```
Comment (palpatim) [inline CHANGELOG.md]: nit: space "`exponential`("

File: AWSAppSyncClient/AWSAppSyncClientConfiguration.swift [hunk 6]
```diff
@@ -16,12 +16,13 @@ public class AWSAppSyncClientConfiguration {
...
     private(set) var store: ApolloStore
     private(set) var subscriptionMetadataCache: AWSSubscriptionMetaDataCache?
-
     private(set) var s3ObjectManager: AWSS3ObjectManager? = nil
     private(set) var presignedURLClient: AWSS3ObjectPresignedURLGenerator? = nil
     private(set) var connectionStateChangeHandler: ConnectionStateChangeHandler? = nil
     private(set) var allowsCellularAccess: Bool = true
     private(set) var autoSubmitOfflineMutations: Bool = true
+    // retry strategy is set during initializing the configuration object and cannot be changed on the fly.
+    let retryStrategy: AWSAppSyncRetryStrategy
```
[the remark this answers is not in the document]
Comment (rohandubal) [reply AWSAppSyncClient/AWSAppSyncClientConfiguration.swift]: Discussed offline; we need to help delta sync determine the retry backoff.

File: AWSAppSyncClient/AWSAppSyncClientConfiguration.swift [hunk 7]
```diff
@@ -63,13 +68,16 @@ public class AWSAppSyncClientConfiguration {
...
+    ///
     public init(url: URL,
                 serviceRegion: AWSRegionType,
                 networkTransport: AWSNetworkTransport,
                 cacheConfiguration: AWSAppSyncCacheConfiguration? = nil,
                 connectionStateChangeHandler: ConnectionStateChangeHandler? = nil,
                 s3ObjectManager: AWSS3ObjectManager? = nil,
-                presignedURLClient: AWSS3ObjectPresignedURLGenerator? = nil) {
+                presignedURLClient: AWSS3ObjectPresignedURLGenerator? = nil,
+                retryStrategy: AWSAppSyncRetryStrategy = .exponential) {
```
[the remark this answers is not in the document]
Comment (rohandubal) [reply AWSAppSyncClient/AWSAppSyncClientConfiguration.swift]: Will remove.

File: AWSAppSyncClient/AWSAppSyncHTTPNetworkTransport.swift (hunk 2 above)
Comment (rohandubal) [reply AWSAppSyncClient/AWSAppSyncHTTPNetworkTransport.swift]: will update.

File: AWSAppSyncClient/AWSAppSyncRetryStrategy.swift [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+//
+
+import Foundation
+
+
+/// The retry strategy to be used by the `AWSAppSyncClient`.
+/// You can specify this in the `AWSAppSyncClientConfiguration`.
+///
+/// - exponential: Backs off exponentially before retrying a HTTP request. Starts from 400ms and grows exponentially w/ jitter; stops the retries after the back off reaches 5 minutes.
+/// - aggressive: Aggressively retries every 1s w/ jitter for upto 12 attempts.
```
[the remark this answers is not in the document]
Comment (rohandubal) [reply AWSAppSyncClient/AWSAppSyncRetryStrategy.swift]: will update type and update number of attempts to 30.

File: CHANGELOG.md (hunk 5 above)
Comment (rohandubal) [reply CHANGELOG.md]: Will udpate

**Pull request merged**